// solar/scene/defaultscene.js

module.exports = function(domContainer) {
    var initNear = 10;
    var initFar = 10000;
    var width = domContainer.width();
    var height = domContainer.height();

    var renderer;
    renderer = new THREE.WebGLRenderer();
    renderer.antialias = true;
    renderer.shadowMapEnabled = true;
    renderer.shadowMapSoft = true;
    renderer.setSize( width, height );
    domContainer.append( renderer.domElement );

    var camera = new THREE.PerspectiveCamera( 65, width/height, initNear, initFar );
    camera.position.y = 250;
    camera.position.z = 400;

    var scene = new THREE.Scene();

    function animate(prerender, render) {
        var last = -1;
        if ( render == undefined ) {
            render = function() {
                renderer.render(scene, camera);
            }
        }
        if ( prerender == undefined ) {
            prerender = function(){}
        }
        var animationLoop = function(timestamp) {
            if ( timestamp > last ) {
                requestAnimationFrame(animationLoop);
                prerender(timestamp);
                render(timestamp);
                last = timestamp;
            }
        };
        animationLoop(0);
    }

    return {
        renderer: renderer,
        camera: camera,
        scene: scene,
        add: function(thingToAdd) {
            scene.add(thingToAdd);
        },
        addLights: function(lights) {
            for ( var i = 0; i < lights.length; i++ ) {
               scene.add(lights[i]);
            }
        },
        animate: animate
    }
};
